nsqd/channel.go

```diff
@@ -303,17 +330,52 @@ func (c *Channel) PutMessage(m *Message) error {
 }
 
 func (c *Channel) put(m *Message) error {
-	select {
-	case c.memoryMsgChan <- m:
-	default:
-		err := writeMessageToBackend(m, c.backend)
-		c.nsqd.SetHealth(err)
-		if err != nil {
-			c.nsqd.logf(LOG_ERROR, "CHANNEL(%s): failed to write message to backend - %s",
-				c.name, err)
-			return err
+	if c.topologyAwareConsumption {
+		// Attempt zone local, region local and finally the memory channel
+		// we do this to ensure that we preferentially deliver messages based on toplogy
+		//
+		// Because messagePump is intermittently unavailable while writing a msg to a client
+		// we continue to have higher priority channels in the select loop, this means at each
```

Comment:
small re-wording for clarity (i hope)

```suggestion
		// Because messagePump is intermittently unavailable while writing a msg to a client
		// we continue to add lower priority channels in the select loop, this means at each
```